### ISSUE TITLE
New Discrete Diffusion Model: Wuerstchen

### ISSUE THREAD

**Issue by patrickvonplaten** [huggingface/diffusers#3706]
### Model/Pipeline/Scheduler description

The amazing @dome272 published a new diffusion model, called Wuerstchen: https://github.com/dome272/Wuerstchen

I think it would be a very nice addition to the library given it's superior speed.

Anybody interested in picking this one up?

### Open source status

- [X] The model implementation is available
- [X] The model weights are available (Only relevant if addition is not a scheduler).

### Provide useful links for the implementation

Paper: https://arxiv.org/abs/2306.00637
Code: https://github.com/dome272/Wuerstchen
Model: https://github.com/dome272/Wuerstchen

@dome272

**Comment by alexblattner** [comment]
this is nice, but have you considered implementing this:
https://github.com/jannerm/ddpo

It's training through human input but improved

**Comment by Abhinay1997** [comment]
Wuerstchen is a fascinating read!

**Comment by kadirnar** [comment]
> this is nice, but have you considered implementing this: https://github.com/jannerm/ddpo
> 
> It's training through human input but improved here's a paper: https://rl-diffusion.github.io/

Are you considering adding it to the Diffusers library?

**Comment by adit299** [comment]
I am new to this repository, but I wouldn't mind taking a crack at it!

**Comment by dg845** [comment]
Not sure which model @adit299 is referring to but I'd be interested in trying to implement the DDPO model if no one else is planning to work on it :).

**Comment by adit299** [comment]
> Not sure which model @adit299 is referring to but I'd be interested in trying to implement the DDPO model if no one else is planning to work on it :).

I was referring to the Wuerstchen model.

**Comment by dome272** [comment]
Would love to if you would give it a shot.

**Comment by dg845** [comment]
Might be a little off-topic but my understanding is that the DDPO models are already supported in `diffusers` from a pipeline perspective. The models released with the original code are all finetuned from Stable Diffusion v1.4 (and so are compatible with `StableDiffusionPipeline`) and already have `diffusers`-compatible checkpoints ([1](https://huggingface.co/kvablack/ddpo-compressibility), [2](https://huggingface.co/kvablack/ddpo-incompressibility), [3](https://huggingface.co/kvablack/ddpo-aesthetic), [4](https://huggingface.co/kvablack/ddpo-alignment)).

Something that could still be added is an example for finetuning a model using DDPO, not sure if the original authors are planning to work on that or not.

**Comment by rudolphyo** [comment]
@dome272 Hi!

I was eager to dive in implementation of Würstchen pipeline to examples/community, but noticed that you are one of the original authors and already making some progress. 
My question is, would you mind if I'll take my chances in implementing, or you'd like to do it yourself? 
Maybe you see a way we can collaborate on this ? 

By the way, paper have gorgeous generation examples!

**Comment by dome272** [comment]
Hey @rudolphyo !
I would to if you would give it a shot. We can also talk more about this on discord. Feel free to add me ```discord=dome1```

**Comment by rudolphyo** [comment]
@dome272, thanks for such fast reply!
I'll definitely circle back to you on discord in case of stumbling upon peculiar issues

**Comment by rudolphyo** [comment]
@patrickvonplaten hi!
I'd like to share a brief plan here and also have some guidance, cause it appeared more complex than I thought at first.

* I was hoping to reuse model's code from diffusers/models (e.g. vq_model.py, resnet.py) as it stated in contribution guidelines but it seems like it would be more explicit and beneficial to add new module for VQ-GAN[0] ? Same applies for Paella which samples from predicted noise ("token predictor") 
* As I can see from other PRs models should be introduced in diffusers/models  so its ok to exceed scope of examples/community? I was able to find some info bit by bit on this matter in various issues and discussions, but maybe I missed some resource and you can point me in the right direction ?  

here you can find links to author's original code
[0] https://github.com/dome272/Wuerstchen/blob/main/vqgan.py
[1] https://github.com/dome272/Wuerstchen/blob/main/modules.py

thanks in advance !

**Comment by patrickvonplaten** [comment]
Hey @rudolphyo,

Thanks for the nice summary! Fully agree with you think a `vq_gan.py` file in `diffusers/models` is warranted here as wuerstchen works with tokens. More than happy to guide you in a PR. Also cc @patil-suraj here

**Comment by dome272** [comment]
This might provide a good starting point as it improves a few things, makes it cleaner and more object-oriented https://github.com/xvdp/Wuerstchen/blob/xdev/wurst_c.py. Thanks @xvdp

**Comment by dome272** [comment]
This might provide a good starting point as it improves a few things, makes it cleaner and more object-oriented https://github.com/xvdp/Wuerstchen/blob/xdev/wurst_c.py. Thanks @xvdp

**Comment by dome272** [comment]
@rudolphyo Do you have any updates if you had the time to start? I would really appreciate if you could add me on discord as we would like to get involved too. Discord: dome1

**Comment by Lime-Cakes** [comment]
Will training code for all three stages (A,B,C) be added?

**Comment by rudolphyo** [comment]
@dome272 Hi, yeah, I got a chance to kick-start implementation, my initial plan was to make both vq-gan and paella be configurable in terms of encoder-decoder backbones (ie making it viable as a go-to model for other pipelines as well, not just wurst) but this intention got me into issues with using original model's weights, so I decided to make a step back here and make it just work with wurst for now. I guess I'll be done by the end of this week with further polishing on the next, (ie adding tests for each of the model we have here). Does it sounds reasonable to you ? I'll start to lay it all out in a day at my fork.
